nav2-usability/scenario_code/lib/sign-in-routing/sign_in_routing_go_router_prototype.dart

```diff
@@ -0,0 +1,226 @@
+// Copyright 2022, the Flutter project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+import 'package:go_router_prototype/go_router_prototype.dart';
+
+void main() {
+  runApp(const BooksApp());
+}
+
+class Credentials {
+  final String username;
+  final String password;
+
+  Credentials(this.username, this.password);
+}
+
+class Authentication extends ChangeNotifier {
+  bool _signedIn = false;
+
+  bool isSignedIn() => _signedIn;
+
+  Future<void> signOut() async {
+    _signedIn = false;
+    notifyListeners();
+  }
+
+  Future<bool> signIn(String username, String password) async {
+    _signedIn = true;
+    notifyListeners();
+    return true;
+  }
+}
+
+class BooksApp extends StatefulWidget {
+  const BooksApp({Key? key}) : super(key: key);
+
+  @override
+  State<StatefulWidget> createState() => _BooksAppState();
+}
+
+class AppState {
+  final Authentication auth;
+
+  AppState(this.auth);
+
+  Future<bool> signIn(String username, String password) async {
+    var success = await auth.signIn(username, password);
+    return success;
+  }
+
+  Future<void> signOut() async {
+    await auth.signOut();
+  }
+}
+
+class _BooksAppState extends State<BooksApp> {
+  final AppState _appState = AppState(Authentication());
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp.router(
+      routeInformationParser: _router.parser,
+      routerDelegate: _router.delegate,
+    );
+  }
+
+  late final _router = GoRouter(
+    refreshListenable: _appState.auth,
+    routes: [
+      ShellRoute(
+        path: '/',
+        builder: (context, child) => child,
+        redirect: (state) async {
+          final signedIn = _appState.auth.isSignedIn();
+          if (!signedIn) return '/signin';
```

Comment:
Do we ignore redirects if the new location is the current location? If yes, are there valid scenarios where a user may want to redirect to the current route - perhaps to force a refresh of some logic? If no, should this logic be updated to not redirect if the current route is `/signin`?

Reply:
Right now, GoRouter displays an ErrorScreen if it detects a redirect loop, including the case where a route redirects to itself:

```dart
    testWidgets('Can redirect to itself', (WidgetTester tester) async {
      final List<GoRoute> routes = <GoRoute>[
        GoRoute(
          path: '/',
          redirect: (_) => '/',
          builder: (BuildContext context, GoRouterState state) =>
              const Page1Screen(),
        ),
      ];

      final GoRouter router = await _router(routes, tester);
      final List<GoRouteMatch> matches = router.routerDelegate.matches;
      log.severe(
          'Exception: ${(router.screenFor(matches.first) as ErrorScreen).ex}');
      expect(router.screenFor(matches.last).runtimeType, Page1Screen);
    });
```

Which prints:

```
[SEVERE] GoRouter tests: Exception: Exception: redirect loop detected: / => /
```

The prototype allows this, which is why this sample works.

I checked the `go_router` version of this sample, and it needed to be fixed: [b5b9a01](https://github.com/flutter/uxr/pull/82/commits/b5b9a01fffeaafe743ae6d3a0c8fb46a83ebb2a8)

Reply:
I filed https://github.com/flutter/flutter/issues/106214 to track this